table/block_based_table_reader.cc

```diff
@@ -1987,10 +1989,24 @@ void BlockBasedTableIterator::InitDataBlock() {
         num_file_reads_++;
         // Do not readahead more than kMaxReadaheadSize.
         readahead_size_ = std::min(kMaxReadaheadSize, readahead_size_);
-        table_->get_rep()->file->Prefetch(data_block_handle.offset(),
-                                          readahead_size_);
-        readahead_limit_ = static_cast<size_t>(data_block_handle.offset()
-          + readahead_size_);
+
+        // Discarding the return status of Prefetch calls intentionally, as we
+        // can fallback to reading from disk if Prefetch fails.
+        if (!rep->file->use_direct_io()) {
+          // buffered i/o
+          rep->file->Prefetch(data_block_handle.offset(), readahead_size_);
+          readahead_limit_ =
+              static_cast<size_t>(data_block_handle.offset() + readahead_size_);
+        } else if (!rep->for_compaction) {
+          // direct i/o -- but not for compaction reads.
+          if (!prefetch_buffer_) {
+            prefetch_buffer_.reset(new FilePrefetchBuffer());
+          }
+          prefetch_buffer_->Prefetch(
+              rep->file.get(), data_block_handle.offset(), readahead_size_);
+          readahead_limit_ =
+              prefetch_buffer_->Offset() + prefetch_buffer_->Length();
```

Comment:
I suggest we move the logic to `prefetch_buffer_` so that it can be reused later. In that case, we may not need to do anything here, and the prefetching decision can be done when TryReadFromCache() is called. Just like what is done in ReadaheadRandomAccessFile::Read(). The benefit is that, this can be more easily reused by partition index prefetching, which goes to a different code path.

Reply:
This is a great suggestion. Done.